### PR TITLE
[full-ci] [tests-only]WebUI tests added to share resource with multiple users at once

### PR DESCRIPTION
**Pull request by grgprarup**
## Description
This PR adds WebUI tests for sharing resources with multiple users at once. Since the CI is green every time but when testing locally we might face the issue mentioned here https://github.com/owncloud/core/issues/40548
Scenarios included:
- `Scenario: user shares file with multiple users at once`
- `Scenario: user shares folder with multiple users at once`
- `Scenario Outline: user shares file with multiple users including non-existing user at once`
- `Scenario: user shares folder with multiple users including non-existing user at once`
- `Scenario: user shares file with multiple users having exact same group name`
- `Scenario: user shares folder with multiple users having exact same group name`

## Related Issue

- Part of https://github.com/owncloud/core/issues/40178 

## How Has This Been Tested?

- Locally
- :robot:

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 1]
```diff
@@ -448,6 +478,21 @@ public function theUserSharesUsingWebUIWithoutClosingDialog(
...
 				);
 			}
+		} elseif ($userOrGroup === "users") {
+			$users = explode(",", $name);
+			$users_array = [];
+			foreach ($users as $user) {
+				if ($this->featureContext->userExists($user)) {
+					$users_array[] = $user;
+				}
+			}
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php:488]: imo, this check only should go while clicking the autocomplete options, but not while typing in the input field

File: tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature [hunk 2]
```diff
@@ -353,3 +353,112 @@ Feature: Sharing files and folders with internal users
...
+      | Carol    |
+    And user "Alice" has uploaded file "filesForUpload/lorem.txt" to "lorem.txt"
+    And user "Alice" has logged in using the webUI
+    When the user shares file "lorem.txt" with users "Brian,Carol" using the webUI
+    And the user re-logs in as "Brian" using the webUI
+    Then file "lorem.txt" should be listed on the webUI
+    And file "lorem.txt" should be marked as shared by "Alice" on the webUI
+    When the user re-logs in as "Carol" using the webUI
+    Then file "lorem.txt" should be listed on the webUI
+    And file "lorem.txt" should be marked as shared by "Alice" on the webUI
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature:372]: api checks could be done to save time.

File: tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature [hunk 3]
```diff
@@ -353,3 +353,112 @@ Feature: Sharing files and folders with internal users
...
+      | Alice    |
+      | Brian    |
+      | Carol    |
+    And user "Alice" has created folder "simple-folder"
+    And user "Alice" has logged in using the webUI
+    When the user shares folder "simple-folder" with users "Brian,Carol" using the webUI
+    And the user re-logs in as "Brian" using the webUI
+    Then folder "simple-folder" should be listed on the webUI
+    And folder "simple-folder" should be marked as shared by "Alice" on the webUI
+    And the user re-logs in as "Carol" using the webUI
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature:387]:
```suggestion
    When the user re-logs in as "Carol" using the webUI
```

File: tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature [hunk 4]
```diff
@@ -353,3 +353,112 @@ Feature: Sharing files and folders with internal users
...
+      | Alice    |
+      | Brian    |
+      | Carol    |
+    And user "Alice" has created folder "simple-folder"
+    And user "Alice" has logged in using the webUI
+    When the user shares folder "simple-folder" with users "Brian,David,Carol" using the webUI
+    And the user re-logs in as "Brian" using the webUI
+    Then folder "simple-folder" should be listed on the webUI
+    And folder "simple-folder" should be marked as shared by "Alice" on the webUI
+    And the user re-logs in as "Carol" using the webUI
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature:426]: when

File: tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature [hunk 5]
```diff
@@ -353,3 +353,112 @@ Feature: Sharing files and folders with internal users
...
+      | Brian    |
+      | Carol    |
+    And group "Brian,Carol" has been created
+    And user "Alice" has created folder "simple-folder"
+    And user "Alice" has logged in using the webUI
+    When the user shares folder "simple-folder" with users "Brian,Carol" using the webUI
+    And the user re-logs in as "Brian" using the webUI
+    Then folder "simple-folder" should be listed on the webUI
+    And folder "simple-folder" should be marked as shared by "Alice" on the webUI
+    And the user re-logs in as "Carol" using the webUI
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature:462]: when

File: tests/acceptance/features/bootstrap/WebUISharingContext.php (hunk 1 above)
Comment (grgprarup) [reply tests/acceptance/features/bootstrap/WebUISharingContext.php:488]: Thanx for the catch. Now, the non-existent user is removed and passed to function as `nameToMatch`.

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 6]
```diff
@@ -448,6 +478,22 @@ public function theUserSharesUsingWebUIWithoutClosingDialog(
...
```
Comment (kiranparajuli589) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php:488]: is this still needed? `$users_array` seem unused.

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 7]
```diff
@@ -174,6 +174,36 @@ public function theUserSharesFileFolderWithUserUsingTheWebUI(
 		);
 	}
 
+	/**
+	 * @When /^the user shares (?:file|folder) "([^"]*)" with users "([^"]*)" using the webUI$/
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php:178]: Could be simpler like this
```suggestion
	 * @When the user shares file/folder :folder with users :users using the webUI
```

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 8]
```diff
@@ -174,6 +174,36 @@ public function theUserSharesFileFolderWithUserUsingTheWebUI(
...
+	 *
+	 * @param string $folder
+	 * @param string $users
+	 * @param int $maxRetries
+	 * @param boolean $quiet
+	 * @param boolean $expectedToWork
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php:184]: unnecessary params?
```suggestion
```

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 9]
```diff
@@ -402,6 +403,33 @@ public function shareWithUser(
...
+	 *
+	 * @return void
+	 * @throws ElementNotFoundException|Exception
+	 */
+	public function shareWithUsers(
+		?string  $name,
+		?string $nameToMatch,
+		Session $session,
+		bool $quiet,
+		int     $maxRetries = 5
```
Comment (saw-jan) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php:422]:
```suggestion
		int $maxRetries = 5
```

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 10]
```diff
@@ -402,6 +403,33 @@ public function shareWithUser(
...
+	 * @param boolean $quiet
+	 * @param int $maxRetries
+	 *
+	 * @return void
+	 * @throws ElementNotFoundException|Exception
+	 */
+	public function shareWithUsers(
+		?string  $name,
+		?string $nameToMatch,
+		Session $session,
```
Comment (saw-jan) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php:420]: Should `$name` be optional? And optional params should come after compulsory ones 
```suggestion
	public function shareWithUsers(
		string  $name,
		string $nameToMatch,
		Session $session,
```

File: tests/acceptance/features/bootstrap/WebUISharingContext.php (hunk 6 above)
Comment (grgprarup) [reply tests/acceptance/features/bootstrap/WebUISharingContext.php:488]: `$users_array` is used to store existing users and later used to create `$nameToMatch`. I tried string concatenation but not suitable for this case.

File: tests/acceptance/features/bootstrap/WebUISharingContext.php (hunk 8 above)
Comment (grgprarup) [reply tests/acceptance/features/bootstrap/WebUISharingContext.php:184]: Removing those param gives phpDocComment warnings.

Comment (saw-jan) [reply tests/acceptance/features/bootstrap/WebUISharingContext.php:184]: Yes, you need to remove them from the functions too

File: tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature [hunk 11]
```diff
@@ -353,3 +353,95 @@ Feature: Sharing files and folders with internal users
...
+      | David,Brian,Carol |
+
+
+  Scenario: user shares file with multiple users having exact same group name
+    Given these users have been created with default attributes and without skeleton files:
+      | username |
+      | Alice    |
+      | Brian    |
+      | Carol    |
+    And group "Brian,Carol" has been created
```
Comment (saw-jan) [inline tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature:428]: Add user to `Brian,Carol` group and share doesn't exist for that user can be checked

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 12]
```diff
@@ -174,6 +174,27 @@ public function theUserSharesFileFolderWithUserUsingTheWebUI(
 		);
 	}
 
+	/**
+	 * @When the user shares file/folder :resource with users :users using the webUI
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php:178]:
```suggestion
	 * @When the user shares (?:file|folder) :resource with users :users using the webUI
```

File: tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature [hunk 13]
```diff
@@ -353,3 +353,95 @@ Feature: Sharing files and folders with internal users
...
+
+  Scenario: user shares folder with multiple users at once
+    Given these users have been created with default attributes and without skeleton files:
+      | username |
+      | Alice    |
+      | Brian    |
+      | Carol    |
+    And user "Alice" has created folder "simple-folder"
+    And user "Alice" has logged in using the webUI
+    When the user shares folder "simple-folder" with users "Brian,Carol" using the webUI
```
Comment (SagarGi) [inline tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature:379]: what if we send like "Brian, Carol" with a space?

File: tests/acceptance/features/bootstrap/WebUISharingContext.php (hunk 12 above)
Comment (saw-jan) [reply tests/acceptance/features/bootstrap/WebUISharingContext.php:178]: that doesn't work with cucumber expression

File: tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature (hunk 13 above)
Comment (grgprarup) [reply tests/acceptance/features/webUISharingInternalUsers1/createShareWithUsers.feature:379]: Spaces won't matter.

**Review by saw-jan — COMMENTED** [review]
LGTM :+1:

**Review by kiranparajuli589 — COMMENTED** [review]
lgtm :+1:

**Review by saw-jan — COMMENTED** [review]
:+1:

**Review by individual-it — APPROVED** [review]

**Review by phil-davis — APPROVED** [review]

**Pull request merged**